rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+
+        It is better to use :py:func:`~.write` which wraps this method and
+        allows writing bytestrings of arbitrary length.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start writing the data.
+        data : :py:class:`bytes`
+            Data to write into memory.  Must be <= 256 bytes.
+        data_type : :py:class:`~rig.machine_controller.consts.DataType`
+            The size of the data to write into memory.
+        """
+        length_bytes = len(data)
+        self.send_scp(x, y, p, SCPCommands.write, address, length_bytes,
+                      int(data_type), data, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def read(self, address, length_bytes, x=Required, y=Required, p=Required):
+        """Read a bytestring from an address in memory.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start reading the data.
+        length_bytes : int
+            The number of bytes to read from memory.
+
+        Returns
+        -------
+        :py:class:`bytes`
+            The data is read back from memory as a bytestring.
+        """
+        # Make calls to the lower level read method until we have read
+        # sufficient bytes.
+        data = b''
+        while len(data) < length_bytes:
+            # Determine the number of bytes to read
+            reads = min(256, length_bytes - len(data))
+
+            # Determine the data type to use
+            dtype = address_length_dtype[(address % 4, reads % 4)]
+
+            # Perform the read and increment the address
+            data += self._read(x, y, p, address, reads, dtype)
+            address += reads
+
+        return data
+
+    def _read(self, x, y, p, address, length_bytes, data_type=DataType.byte):
+        """Read a bytestring from an address in memory.
+
+        It is better to use :py:func:`~.read` which wraps this method and
+        allows reading bytestrings of arbitrary length.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start reading the data.
+        length_bytes : int
+            The number of bytes to read from memory, must be <= 256.
+        data_type : DataType
+            The size of the data to write into memory.
+
+        Returns
+        -------
+        :py:class:`bytes`
+            The data is read back from memory as a bytestring.
+        """
+        read_scp = self.send_scp(x, y, p, SCPCommands.read, address,
+                                 length_bytes, int(data_type),
+                                 expected_args=0)
+        return read_scp.data
+
+    @ContextMixin.use_contextual_arguments
+    def iptag_set(self, iptag, addr, port, x=Required, y=Required):
+        """Set the value of an IPTag.
+
+        Parameters
+        ----------
+        iptag : int
+            Index of the IPTag to set
```

Comment:
Done (aside for x and y which I think can be documented once in the class docstring?)